diff --git a/common/r35/tegraflash/tegrasign_v3_util.py b/common/r35/tegraflash/tegrasign_v3_util.py
index f341ec9..9675ec4 100755
--- a/common/r35/tegraflash/tegrasign_v3_util.py
+++ b/common/r35/tegraflash/tegrasign_v3_util.py
@@ -19,7 +19,10 @@ import subprocess
 import re
 import time
 import traceback
-import yaml
+try:
+    import yaml
+except ImportError:
+    print("WARNING: failed to import yaml")
 
 AES_128_HASH_BLOCK_LEN = 16
 AES_256_HASH_BLOCK_LEN = 16
@@ -178,8 +181,11 @@ class KDF:
         self.bootmode = None
 
     def parse_file(self, p_key, kdf_file, internal = None):
-        with open(kdf_file) as f:
-            params = yaml.safe_load(f)
+        try:
+            with open(kdf_file) as f:
+                params = yaml.safe_load(f)
+        except ImportError:
+            print("WARNING: kdf: yaml unavailable")
 
         tokens = {'IV':'--iv', 'AAD':'--aad', 'VER':None, 'DERSTR':None, 'CHIPID':None,
                   'MAGICID':None, 'FLAG':None, 'BL_DERSTR':None, 'FW_DERSTR':None}
