From e4dace0534fc80aa8e064e4217e57b1fee1116f6 Mon Sep 17 00:00:00 2001
From: Aaron Kling <webgeek1234@gmail.com>
Date: Mon, 16 Jun 2025 16:06:59 -0500
Subject: [PATCH] common: Update tnspec for py3

---
 common/rel-24/tegraflash/tnspec.py | 113 ++++++++++++++---------------
 1 file changed, 55 insertions(+), 58 deletions(-)

diff --git a/common/rel-24/tegraflash/tnspec.py b/common/rel-24/tegraflash/tnspec.py
index d41cdbd..72276b0 100755
--- a/common/rel-24/tegraflash/tnspec.py
+++ b/common/rel-24/tegraflash/tnspec.py
@@ -13,7 +13,7 @@
 #
 # NOTE: every feature implemented here must have a correponding test
 
-from __future__ import print_function
+
 from struct import pack, pack_into, unpack, unpack_from, calcsize as csz
 import sys
 import os
@@ -163,7 +163,7 @@ def qopen(fobj, mode='rb'):
         mode - Passed to open() function. N/A when sys.stdout
 
     """
-    if isinstance(fobj, types.StringTypes):
+    if isinstance(fobj, (str,)):
         f = open(fobj, mode)
     elif fobj in [sys.stdout, sys.stdin]:
         f = fobj
@@ -172,7 +172,7 @@ def qopen(fobj, mode='rb'):
     try:
         yield f
     finally:
-        if isinstance(fobj, types.StringTypes):
+        if isinstance(fobj, (str,)):
             f.close()
 
 ###############################################################################
@@ -252,9 +252,8 @@ class CmdNCT(Command):
         out = ''
         if len(key) and key in ['all', 'tnspec'] + [e['name'] for e in NCT.base_entries]:
             if key == 'all':
-                out += '  |      HEADER| %s\n' % str(map(lambda x:
-                                                    x if isinstance(x,str)
-                                                    else hex(x), nct.hdr['val']))
+                out += '  |      HEADER| %s\n' % str([x if isinstance(x,str)
+                                                    else hex(x) for x in nct.hdr['val']])
                 out += '  |      TNSPEC| %s\n' % (
                         ('[%X] ' % nct.hdr['val'][8] +
                          '- use "tnspec nct dump tnspec" to view')
@@ -338,7 +337,7 @@ class CmdNCT(Command):
 
         # remove reserved keys
         for k in reserved_keys:
-            if k in _hwspec.keys():
+            if k in list(_hwspec.keys()):
                 pr_dbg("Removing reserved keys", k)
                 del _hwspec[k]
 
@@ -587,12 +586,12 @@ class CmdSpec(Command):
 
         q = spec.get_specs(query)
 
-        if isinstance(q, types.StringTypes):
+        if isinstance(q, (str,)):
             pr(1, q)
-        elif isinstance(q, types.BooleanType):
+        elif isinstance(q, bool):
             pr(1, repr(q).lower())
-        elif isinstance(q, types.ListType):
-            pr(1, '\n'.join(map(lambda x: str(x), q)))
+        elif isinstance(q, list):
+            pr(1, '\n'.join([str(x) for x in q]))
         elif q == None or not len(q):
             pass
         else:
@@ -655,28 +654,28 @@ class NCT(object):
         if e['val'] != None:
             # handle special cases first
             if e['name'] == 'board_info':
-                return 'Proc: ' + '.'.join(map(lambda x: str(x),e['val'][:3])) + \
-                       ' PMU: ' + '.'.join(map(lambda x: str(x), e['val'][3:6])) + \
-                       ' Disp: ' + '.'.join(map(lambda x: str(x), e['val'][-3:]))
+                return 'Proc: ' + '.'.join([str(x) for x in e['val'][:3]]) + \
+                       ' PMU: ' + '.'.join([str(x) for x in e['val'][3:6]]) + \
+                       ' Disp: ' + '.'.join([str(x) for x in e['val'][-3:]])
             elif e['name'] == 'uuid':
                 uuids = bytearray(csz(e['fmt']))
                 # pack flat
                 pack_into(e['fmt'], uuids, 0, *e['val'])
                 # re-unpack
                 uuids_fmt = '64s' * 16
-                uuids = unpack_from(uuids_fmt, buffer(uuids))
+                uuids = unpack_from(uuids_fmt, memoryview(uuids))
                 uuids = [ x.rstrip('\0') for x in uuids ]
                 uuids = [ u for u in uuids if len(u) ]
                 return uuids
 
             elif e['fmt'] == '6B':
-                return ':'.join(map(lambda x: '%02X' % x,e['val']))
+                return ':'.join(['%02X' % x for x in e['val']])
             elif e['fmt'].endswith('s'):
                 return  str(e['val'][0]).rstrip('\0')
             elif e['fmt'].startswith(('I','H')):
                 return e['val'][0]
             else:
-                return map(lambda x: '0x%08x' % x, e['val'])
+                return ['0x%08x' % x for x in e['val']]
 
     @staticmethod
     def _nct_txt_tag(e):
@@ -722,7 +721,6 @@ class NCT(object):
             # use bytearray to make it mutable
             data = bytearray(csz(e['fmt']))
             data[:len(s[name])] = s[name].encode('ascii')
-            data = str(data)
             e['val'] = unpack_from(e['fmt'], data)
 
     def _spec_null(e, s, unused):
@@ -734,7 +732,7 @@ class NCT(object):
         """
         if name in s:
             mac = s[name].split(':')
-            e['val'] = tuple(map(lambda x: int(x, 16), mac))
+            e['val'] = tuple([int(x, 16) for x in mac])
 
     def _spec_board(e, s, name):
         """
@@ -779,7 +777,7 @@ class NCT(object):
                 pr_err("Number of elements (%d) in %s is greater than %d." %
                         (len(s[name]), name, num))
                 sys.exit(1)
-            data[:len(s[name])] = map(lambda x: int(x,0), s[name])
+            data[:len(s[name])] = [int(x,0) for x in s[name]]
             e['val'] = tuple(data)
 
     def _spec_uuid(e, s, *unused_for_now):
@@ -792,7 +790,7 @@ class NCT(object):
         data = bytearray(csz(e['fmt']))
         use_tags = s.get('uuid',{}).get('use_tags', False)
         idx = 0
-        for key, module in s.items():
+        for key, module in list(s.items()):
             if isinstance(module, dict) and 'uuid' in module:
                 if idx >= uuid_max_entries:
                     pr_err("Error: %s.uuid can't be added (only up to %d)" %
@@ -809,7 +807,6 @@ class NCT(object):
                 data[idx*csz(uuid_fmt):len(uuid)] = uuid.encode('ascii')
                 idx += 1
         if idx:
-            data = buffer(data)
             e['val'] = unpack_from(e['fmt'], data)
 
     def _spec_tnspec_id(e, s, *unused):
@@ -829,7 +826,6 @@ class NCT(object):
 
         data = bytearray(csz(e['fmt']))
         data[:len(spec_str)] = spec_str.encode('ascii')
-        data = str(data)
         e['val'] = unpack_from(e['fmt'], data)
 
     # various formats
@@ -970,7 +966,7 @@ class NCT(object):
 
         # read in the header
         self.hdr['val'] = unpack_from(self.hdr['fmt'], self.rawbin)
-        pr_dbg(map(lambda x: x if isinstance(x,str) else hex(x), self.hdr['val']))
+        pr_dbg([x if isinstance(x,str) else hex(x) for x in self.hdr['val']])
 
         self._load_tnspec_from_nct()
 
@@ -1074,16 +1070,16 @@ class NCT(object):
                 sys.exit(1)
 
             # Zero out TNSPEC space
-            self.rawbin[NCT.tnsoffset:NCT.nctbin_size] = '\x00' * (NCT.nctbin_size - NCT.tnsoffset)
+            self.rawbin[NCT.tnsoffset:NCT.nctbin_size] = bytes('\x00' * (NCT.nctbin_size - NCT.tnsoffset), 'utf-8')
             # Write TNSPEC str
-            self.rawbin[NCT.tnsoffset:NCT.tnsoffset + tnspec_length] = tnspec_str
+            self.rawbin[NCT.tnsoffset:NCT.tnsoffset + tnspec_length] = bytes(tnspec_str, 'utf-8')
 
             # Update header information
             hdr = list(self.hdr['val'])
             # Set tnspec length to header
             hdr[7] = tnspec_length
             # Update crc32
-            hdr[8] = zlib.crc32(tnspec_str) & 0xffffffff
+            hdr[8] = zlib.crc32(bytes(tnspec_str, 'utf-8')) & 0xffffffff
             # Convert it back to tuple type
             self.hdr['val'] = tuple(hdr)
             pr_dbg("TNSPEC dump")
@@ -1106,7 +1102,7 @@ class NCT(object):
 
     def _get_exclude(self):
         no_export = self.nctspec.get_specs('__nct__.exclude')
-        if isinstance(no_export, types.ListType):
+        if isinstance(no_export, list):
             return no_export
         else:
             return []
@@ -1228,7 +1224,8 @@ class NCT(object):
 
         # Update TNSPEC Magic
         hdr = list(self.hdr['val'])
-        hdr[5] = NCT.tnspec_versions[self._get_tnspec_version()]['magic']
+        hdr[0] = self.magic.encode('utf-8')
+        hdr[5] = NCT.tnspec_versions[self._get_tnspec_version()]['magic'].encode('utf-8')
         self.hdr['val'] = tuple(hdr)
 
         pack_into(self.hdr['fmt'], rawbin, 0, *self.hdr['val'])
@@ -1264,7 +1261,7 @@ class NCT(object):
 
     # debug only
     def dump(self):
-        pr(1, '  |      HEADER|', map(lambda x: x if isinstance(x,str) else hex(x), self.hdr['val']))
+        pr(1, '  |      HEADER|', [x if isinstance(x,str) else hex(x) for x in self.hdr['val']])
         for e in self.entries:
             if e['val'] != None:
                 pr(1, '%2d|%12s|' % (e['idx'], e['name'][:12]), NCT.print(e))
@@ -1343,7 +1340,7 @@ class Spec(object):
 
         if self.full_spec:
             if self.group in self.src:
-                for s in self.src[self.group].keys():
+                for s in list(self.src[self.group].keys()):
                     pr_dbg('processing ', s)
                     self._flatten(self.group, s)
                 self.specs = self.src[self.group]
@@ -1493,11 +1490,11 @@ class Spec(object):
         """
         Merges two dict objects
         """
-        for k,v2 in new.items():
+        for k,v2 in list(new.items()):
             v1 = base.get(k) # returns None if v1 has no value for this key
 
-            if (isinstance(v1, collections.Mapping) and
-                isinstance(v2, collections.Mapping)):
+            if (isinstance(v1, collections.abc.Mapping) and
+                isinstance(v2, collections.abc.Mapping)):
                 Spec.merge_raw(v1, v2)
             elif k.endswith(tuple(Spec.ops)):
                 Spec._handle_ops(base, new, k)
@@ -1527,13 +1524,13 @@ class Spec(object):
         """
         Removes keys with 'prefix' or keys in 'keys_remove'
         """
-        if not isinstance(d, collections.Mapping): return
-        keys = d.keys()
+        if not isinstance(d, collections.abc.Mapping): return
+        keys = list(d.keys())
         for k in keys:
             _path = Spec._append_path(path, k)
             if k.startswith(prefix) or _path in keys_remove:
                 del d[k]
-            elif isinstance(d[k], collections.Mapping):
+            elif isinstance(d[k], collections.abc.Mapping):
                 Spec._remove_keys(d[k], prefix, keys_remove, _path)
 
     @staticmethod
@@ -1541,14 +1538,14 @@ class Spec(object):
         """
         Removes keys whose value matches 's'
         """
-        if not isinstance(d, collections.Mapping): return
-        keys = d.keys()
+        if not isinstance(d, collections.abc.Mapping): return
+        keys = list(d.keys())
         for k in keys:
             _path = Spec._append_path(path, k)
             if Spec.is_reference(_path): continue
             if d[k] == s:
                 del d[k]
-            elif isinstance(d[k], collections.Mapping):
+            elif isinstance(d[k], collections.abc.Mapping):
                 Spec._remove_vals(d[k], s, _path)
 
     @staticmethod
@@ -1558,9 +1555,9 @@ class Spec(object):
 
         @file://<file>:<key_path> - Replace this key with the contents of <file>
         """
-        if not isinstance(s, collections.Mapping): return
-        for k,v in s.items():
-            if isinstance(v, types.StringTypes):
+        if not isinstance(s, collections.abc.Mapping): return
+        for k,v in list(s.items()):
+            if isinstance(v, (str,)):
                 r = re.match('@file://([\w+-\.]+)', v)
                 if r:
                     ext_spec = r.group(1)
@@ -1568,7 +1565,7 @@ class Spec(object):
                     ext_keys = r.group(1) if r else ''
                     # do a "raw-import"
                     s[k] = Spec(filename=os.path.join(dirname, ext_spec), skip_process=True).get_specs(ext_keys)
-            elif isinstance(v, collections.Mapping):
+            elif isinstance(v, collections.abc.Mapping):
                 Spec._import(s[k], dirname)
 
     @staticmethod
@@ -1590,8 +1587,8 @@ class Spec(object):
                     key = key[:-len(s)]
             return key
 
-        if not isinstance(d, collections.Mapping): return
-        keys = d.keys()
+        if not isinstance(d, collections.abc.Mapping): return
+        keys = list(d.keys())
         for key in keys:
             _path = Spec._append_path(path, key)
             if Spec.is_reference(_path): continue
@@ -1605,7 +1602,7 @@ class Spec(object):
                 d[k] = d[key]
                 del d[key]
                 key = k
-            if isinstance(d[key], collections.Mapping):
+            if isinstance(d[key], collections.abc.Mapping):
                 Spec._normalize(d[key], prefixes, suffixes, _path)
 
     @staticmethod
@@ -1630,7 +1627,7 @@ class Spec(object):
             key = h['key']
 
             if self.full_spec:
-                for k,spec in self.specs.items():
+                for k,spec in list(self.specs.items()):
                     if Spec.is_reference(k): continue
                     if key in spec:
                         spec[key] = h['fn'](spec[key])
@@ -1747,7 +1744,7 @@ class Spec(object):
         if query:
             cur_path = '<ROOT>'
             for k in query.split('.'):
-                if not isinstance(s, collections.Mapping) :
+                if not isinstance(s, collections.abc.Mapping) :
                     pr_err("'%s' must be 'Object Type' to satisfy the query '%s'" %
                             (cur_path, query))
                     pr_err("'%s' gives:" % cur_path)
@@ -1757,7 +1754,7 @@ class Spec(object):
                 s = s.get(k,{})
 
         if keys_only:
-            return s.keys() if isinstance(s, collections.Mapping) else []
+            return list(s.keys()) if isinstance(s, collections.abc.Mapping) else []
         else:
             # return as-is
             return s
@@ -1804,12 +1801,12 @@ class Spec(object):
         # path must not be null
         assert path
         # consider supporting non-string types: dict, val, etc..
-        assert isinstance(val, types.StringTypes)
+        assert isinstance(val, (str,))
 
         path = path.split('.')
         for k in path[:-1]:
             pr_dbg(type(base), k)
-            if (k not in base) or (not isinstance(base[k], collections.Mapping)):
+            if (k not in base) or (not isinstance(base[k], collections.abc.Mapping)):
                 base[k] = {}
             base = base.get(k) # base[k]
 
@@ -1909,7 +1906,7 @@ class SwSpec(Spec):
         # Handle SW spec (implicit inheritance from default, compatible)
 
         if self.full_spec:
-            for spec_id in self.specs.keys():
+            for spec_id in list(self.specs.keys()):
                 # skip SW-specific processing if it's a reference
                 if Spec.is_reference(spec_id): continue
                 local_keys = self.specs[spec_id].get('.',[])
@@ -1927,7 +1924,7 @@ class SwSpec(Spec):
 
                 pr_dbg('tnspec id ', spec_id)
                 for cfg in self.specs[spec_id]:
-                    if not isinstance(self.specs[spec_id][cfg], collections.Mapping): continue
+                    if not isinstance(self.specs[spec_id][cfg], collections.abc.Mapping): continue
                     if cfg in self.reserved_keys_sw: continue
                     if cfg == 'default': continue
                     cfg_spec = copy.deepcopy(spec_default)
@@ -1948,11 +1945,11 @@ class SwSpec(Spec):
         default = options.get('default',True)
         if options.get('apply_override', default):
             if self.full_spec:
-                for spec_id in self.specs.keys():
+                for spec_id in list(self.specs.keys()):
                     if Spec.is_reference(spec_id): continue
                     for cfg in self.specs[spec_id]:
                         pr_dbg('sw.process -> specid:',spec_id, 'cfg:',cfg)
-                        if not isinstance(self.specs[spec_id][cfg], collections.Mapping): continue
+                        if not isinstance(self.specs[spec_id][cfg], collections.abc.Mapping): continue
                         if cfg in self.reserved_keys_sw: continue
                         self.apply_override(spec_id + '.' + cfg, True)
             else:
@@ -1967,7 +1964,7 @@ class HwSpec(Spec):
     HW Spec Class (derived fom Spec)
     """
     def _mac_generator(s):
-        if not isinstance(s, collections.Mapping):
+        if not isinstance(s, collections.abc.Mapping):
             return s
         prefix = s.get('prefix')
         if prefix == None or len(prefix.split(':')) != 3:
@@ -2011,7 +2008,7 @@ class HwSpec(Spec):
         if options.get('apply_override', default):
             # add id and config fields if they are not found
             if self.full_spec:
-                for e in self.specs.keys():
+                for e in list(self.specs.keys()):
                     if Spec.is_reference(e):
                         continue
                     if 'id' not in self.specs[e]:
-- 
2.49.0

